app/lib/features/space/widgets/top_nav.dart

```diff
@@ -26,110 +23,192 @@ class TopNavBar extends ConsumerStatefulWidget {
-      recentWatchScreenTabStateTestProvider;
-
   @override
   void initState() {
     super.initState();
     spaceOverviewTutorials(context: context);
-    recentWatchScreenTabStateTestProvider = FutureProvider.autoDispose
-        .family<TabsState, BuildContext>((ref, context) async {
-      final tabs = await ref.watch(tabsProvider(widget.spaceId).future);
-      final selectedIndex = ref.watch(selectedTabIdxProvider(widget.spaceId));
-      final ctrl = TabController(
-        length: tabs.length,
-        vsync: this,
-        initialIndex: selectedIndex.valueOrNull ?? 0,
-      );
-      ref.onDispose(() {
-        ctrl.dispose(); // we need to clean this up.
-      });
-      return TabsState(tabs, ctrl);
-    });
   }
 
   @override
   Widget build(BuildContext context) {
-    ref.watch(selectedTabIdxProvider(widget.spaceId));
-    final tabState = ref.watch(recentWatchScreenTabStateTestProvider(context));
+    final tabState = ref.watch(tabsProvider(widget.spaceId));
     return tabState.when(
-      data: (tabState) {
-        final tabController = tabState.ctrl;
-        final tabs = tabState.tabs;
+      skipLoadingOnReload: true,
+      data: (tabState) => buildTabs(context, tabState),
+      error: (error, stack) {
+        _log.severe('Error generating space tabs menu', error, stack);
+        return buildError(context, error.toString());
+      },
+      loading: () => buildLoading(context),
+    );
+  }
 
-        return LayoutBuilder(
-          key: spaceOverviewKey,
-          builder: (context, constraints) {
-            final useCols = constraints.maxWidth < (150 * tabs.length);
-            int minItemWidth = useCols ? 90 : 150;
-            final minTotalWidth = minItemWidth * tabs.length;
-            bool scrollBar = false;
-            if (minTotalWidth > constraints.maxWidth) {
-              scrollBar = true;
-            }
-            return Container(
-              width: MediaQuery.of(context).size.width,
-              decoration: BoxDecoration(
-                borderRadius: const BorderRadius.only(
-                  topLeft: Radius.circular(10),
-                  topRight: Radius.circular(10),
-                ),
-                color: Theme.of(context).colorScheme.background,
-              ),
-              margin: const EdgeInsets.symmetric(horizontal: 20, vertical: 10),
-              child: TabBar(
-                controller: tabController,
-                isScrollable: scrollBar,
-                labelStyle: useCols
-                    ? Theme.of(context).textTheme.labelSmall
-                    : Theme.of(context).textTheme.bodySmall,
-                labelColor: Colors.white,
-                onTap: (index) {
-                  final target = tabs[index].target;
-                  context.goNamed(
-                    target,
-                    pathParameters: {'spaceId': widget.spaceId},
+  Widget buildContainer(BuildContext context, Widget child) {
```

Comment:
I understand its usage but this seems like generic component which can be extended across other UI loading/error states, can we put this under ```common``` section if there's a possibility of its reusability?

Reply:
I did that in the past and now we have many of those that are effectively used once ... I decided to change strategy and only generalize within a component for now and only later look up if they is common grounds to generalize further. for now, I'd like to keep it this way.